<a href="https://colab.research.google.com/github/natelson/python/blob/main/machine_learning/spell_checker_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Demo of Spell Checker with NLP (Natural Language Processing)



> The idea here is show some functions the Library of Python [nltk](https://www.nltk.org/api/nltk.html). For this example, I created a litle demo of spell checker. The main idea is to go through the following steps:

> * Read a database with a lot of words, this database is a source of training to 
learn words.
* Normalize these words.
* Create the processing of tokenization of these words.
* Create functions that analyze the best answer.




In [121]:
import requests

In [122]:
def get_text_from_a_url_file(target_url :str):
  response = requests.get(target_url)
  data = response.text
  return data

## Read our database

> When we work in NLP, our database is known as a corpus. That is, it is a body composed of several texts, and each one corresponds to an article on our blog, called a document.

> Therefore, the corpus is a set of documents in Natural Language Processing, and in our case, it is composed of blog articles that form the articles.txt database. Then we will read it in the notebook.


In [123]:
target_url = 'https://raw.githubusercontent.com/natelson/python/main/machine_learning/data/articles.txt'
articles = get_text_from_a_url_file(target_url)

## Size of Vocabulary

> Let's imagine that two students are learning a new language, Portuguese.

> At the end of a year of studies, the first read only a 100-page book, while the other read the entire Game of Thrones collection in Portuguese, with each part having almost a thousand pages. In this case, the second person has much more information about the language, as they have access to a much larger vocabulary, including being able to make corrections.

> Therefore, the number of words is interesting for this learning, which also applies to our spell checker.

> To know if our vocabulary is ideal for the job, we must know how many words it has. After all, a text with ten words makes at most ten corrections, for example. So we will need to have a corpus with a large volume of terms.

> In this case, using only the len() function receiving articles will not help us, as it will only return the number of characters present in the vocabulary, and not the number of words we want.


In [124]:
len(articles)

2605046

## Corpus

> A corpus is a collection of authentic text or audio organized into datasets. Authentic here means text written or audio spoken by a native of the language or dialect. A corpus can be made up of everything from newspapers, novels, recipes, radio broadcasts to television shows, movies, and tweets

> In our case, our corpus comes from articles in Portuguese that are in our articles.txt database


## String Tokenization

> String tokenization is a process where a string is broken into several parts. Each part is called a token. For example, if “I am going” is a string, the discrete parts—such as “I”, “am”, and “going”—are the tokens.

> The process of transforming a text file into small tokens is called Tokenization, which is recurrent in the pre-processing or statistical analysis of textual data. So, as it is widely used, we have tools that facilitate its implementation.

> A well-known one in the field of NLP is the nltk https://www.nltk.org/ or Natural Language Tool Kit, which is a set of tools that implements several methods and algorithms for textual analysis.

> Let's import the nltk package and use the tokenize class, which has tokenization methods implemented, such as word_tokenize() which we will call next. We will also need the punkt package as shown below.

In [125]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

## Create a token of our database

> For this example, we are going to use the word_tokenize function which teturn a tokenized copy of text, using NLTK’s recommended word tokenizer. https://www.geeksforgeeks.org/python-nltk-nltk-tokenizer-word_tokenize/

In [126]:
tokens = nltk.tokenize.word_tokenize(articles, language='portuguese')
print(f'Size of tokens before only_words : {len(tokens)}')




Size of tokens before only_words : 490809


> Our token has a size of 490809, however it has other items that are not part of my vocabulary, punctuation for example. So we need to filter our token so that it only has alphanumeric words. And we implement this in the function below choose_only_words

In [127]:
def choose_only_words(token_list):
  token_words = []
  for token in token_list:
    if token.isalpha():
      token_words.append(token)

  return token_words

In [128]:
tokens_only_words = choose_only_words(tokens)
print(f'Size of tokens after only_words : {len(tokens_only_words)}')

Size of tokens after only_words : 393935


## How many words are in my corpus list?



> The total value of 393935 words previously returned in our corpus does not represent the amount of unique words that can be corrected, after all many of them will appear *repeated *in the texts.

> Therefore, we will need to calculate how many unique words there are without repetition.

> Therefore, we will have to normalize the corpus and turn all the text into lowercase.

> We will define the normalize_token() function with def, then we will pass the word_list and return the normalized_list, which will only be composed of words whose characters have all been converted to lowercase. This will be done with word.lower() being the parameter of .append() of the normalized_list.


In [129]:
def normalize_token(token_list : list):
  for index in range(len(token_list)):
    token_list[index] = str(token_list[index]).lower()
  
  return token_list


In [130]:
tokens_only_words = normalize_token(tokens_only_words)

In [131]:
print(tokens_only_words[:50])

['imagem', 'temos', 'a', 'seguinte', 'classe', 'que', 'representa', 'um', 'usuário', 'no', 'nosso', 'sistema', 'java', 'para', 'salvar', 'um', 'novo', 'usuário', 'várias', 'validações', 'são', 'feitas', 'como', 'por', 'exemplo', 'ver', 'se', 'o', 'nome', 'só', 'contém', 'letras', 'cpf', 'só', 'e', 'ver', 'se', 'o', 'usuário', 'possui', 'no', 'mínimo', 'anos', 'veja', 'o', 'método', 'que', 'faz', 'essa', 'validação']



> Now that we have our normalized list, we can remove word repetitions, and there are a few ways to do that.

> We can either build an algorithm or work with more interesting ideas in Python, as it has methods that implement the idea of mathematical sets, such as set().

> With that, we pass our normalized list to set() in the next cell, removing the repetitions.



In [132]:
tokens_only_words_unique = set(tokens_only_words)

In [133]:
print(f'Size of tokens after normalize and remove duplicates : {len(tokens_only_words_unique)}')

Size of tokens after normalize and remove duplicates : 17653


> Running the code, we'll get 17653 of different types, which is quite an interesting value.

> This will be the number of words that our spell checker will learn to correct, remembering that we have a total of 393,935 words in the corpus.

> According to t[his link with a Duolingo article](https://www.optilingo.com/blog/general/how-many-words-do-you-need-to-know-to-become-fluent-in-a-language/), you need to know about 800 to 1,000 words of a language to be at a basic level, 1,500 to 2,000 for intermediate, and 3,000 to 4,000 words to be fluent.

> Therefore, as our algorithm will learn to correct more than 17,000 questions, it is more than enough for us to build a good spell checker, without taking into account the variations between words and other details that we will see later.

## What if I spelled it wrong?
> We want our spell checker to get the term "logic" misspelled and return it corrected as "logic".

> We will discover the correct word because it is among the more than 17,000 words known by counting previously.

> However, we have not yet implemented the transition from "lgica" to "lógica" in fact.

> We will take the misspelled word and generate new words, where one of them could be the correct one. In our specific problem, we have the letter "ó" less.

> Our NLP model will need an algorithm capable of inserting an extra letter into this misspelled term. The algorithm below solves this.

>For this, we need to take the typed word and generate numerous variations of it to see which one is correct, so we can insert a missing letter, invert a letter or even remove a letter.

> These implementations follow below

> Take a list of one-word tuples and insert a letter into each of the list and return, example of passed list [('','lgica'), ('l','gica'), ('lg','ica'), ..] and returned list [('a','lgica'), ('la','gica'), ('lga','ica'), ..].

In [134]:
def concatenate_letter_with_tuple_list(list_tuples_word :list):
  new_words = []
  letters = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
  for right, left in list_tuples_word:
    for letter in letters:
      new_words.append(right + letter + left)

  return new_words
  





> Take a list of one-word tuples and swap a letter into each of the list and return, example of passed list [('','lgica'), ('l','gica'), ('lg','ica'), ..] and returned list [('g','lica'), ('li','gca'), ('lgc','ia'), ..].

In [135]:
def change_letter_with_tuple_list(list_tuples_word :list):
  new_words = []
  letters = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
  for right, left in list_tuples_word:
    for letter in letters:
      new_words.append(right + letter + left[1:])

  return new_words

> Take a list of one-word tuples and remove a letter into each of the list and return, example of passed list [('','lgica'), ('l','gica'), ('lg','ica'), ..] and returned list [('','gica'), ('l','ica'), ('lg','ca'), ..].

In [136]:
def remove_letter_with_tuple_list(list_tuples_word :list):
  new_words = []
  for right, left in list_tuples_word:
      new_words.append(right + left[1:])

  return new_words

In [137]:
def reverse_letter_with_tuple_list(list_tuples_word :list):
  new_words = []
  for right, left in list_tuples_word:
      if len(left) > 1:
        new_words.append(right + left[1]+ left[0] + left[2:])

  return new_words

In [138]:
def word_variations(word_base :str):
  list_tuples_word = []
  for i in range(len(word_base) +1):
    list_tuples_word.append((word_base[:i], word_base[i:]))
  
  word_variotions_list =  concatenate_letter_with_tuple_list(list_tuples_word)
  word_variotions_list +=  remove_letter_with_tuple_list(list_tuples_word)
  word_variotions_list +=  change_letter_with_tuple_list(list_tuples_word)
  word_variotions_list +=  reverse_letter_with_tuple_list(list_tuples_word)
  return word_variotions_list


In [139]:
word_variations_list = word_variations('ólgica')
print(word_variations_list)


['aólgica', 'bólgica', 'cólgica', 'dólgica', 'eólgica', 'fólgica', 'gólgica', 'hólgica', 'iólgica', 'jólgica', 'kólgica', 'lólgica', 'mólgica', 'nólgica', 'oólgica', 'pólgica', 'qólgica', 'rólgica', 'sólgica', 'tólgica', 'uólgica', 'vólgica', 'wólgica', 'xólgica', 'yólgica', 'zólgica', 'àólgica', 'áólgica', 'âólgica', 'ãólgica', 'èólgica', 'éólgica', 'êólgica', 'ìólgica', 'íólgica', 'îólgica', 'òólgica', 'óólgica', 'ôólgica', 'õólgica', 'ùólgica', 'úólgica', 'ûólgica', 'çólgica', 'óalgica', 'óblgica', 'óclgica', 'ódlgica', 'óelgica', 'óflgica', 'óglgica', 'óhlgica', 'óilgica', 'ójlgica', 'óklgica', 'óllgica', 'ómlgica', 'ónlgica', 'óolgica', 'óplgica', 'óqlgica', 'órlgica', 'óslgica', 'ótlgica', 'óulgica', 'óvlgica', 'ówlgica', 'óxlgica', 'óylgica', 'ózlgica', 'óàlgica', 'óálgica', 'óâlgica', 'óãlgica', 'óèlgica', 'óélgica', 'óêlgica', 'óìlgica', 'óílgica', 'óîlgica', 'óòlgica', 'óólgica', 'óôlgica', 'óõlgica', 'óùlgica', 'óúlgica', 'óûlgica', 'óçlgica', 'ólagica', 'ólbgica', 'ólcgica'

## Did you mean ?
> Now that we can take the word typed, generating several variations of it to handle cases of typos, how can we look at our corpus and know which word is most likely to be correct?

> In this step, we will calculate the probability of the generated results.

> To build the word_max_probability() function in a new cell, we will first need to do this calculation with a word; the frequency of the term “logic” will be the number of times it appears in the corpus in relation to the total number of words. To calculate, nltk has the .FreqDist() function.

> For this, we will create a function called list_frequency_words being equal to the method that calculates the Frequency Distribution of words, and we will pass the normalized_list as a parameter

In [140]:
def list_frequency_words(word_list):
  return nltk.FreqDist(word_list)

In [141]:
words_frequency = list_frequency_words(tokens_only_words)
print(words_frequency["lógica"])
total_words = len(tokens_only_words)

87


> In the word_max_probability() function, we will calculate the probability of a word appearing in the corpus.

In [142]:
def word_max_probability(word :str):
  word_frequency = words_frequency[word]/total_words
  return word_frequency

print(round(word_max_probability('lógica'),5))

0.00022


## Finally the spelling correct
 > In the spell_checker function, we pass a right or wrong word and return the most likely word that the user meant. Then the function generates the list of possible variations of the word and with the word_max_probability function checks which one is more likely to be correct.

In [143]:
def spell_checker(word :str):
  word_variations_list = word_variations(word)
  word_correct = max(word_variations_list, key=word_max_probability)
  return word_correct


In [144]:
print(f'Você quis dizer {spell_checker("llógica")} ?')

Você quis dizer lógica ?


In [145]:
print(f'Você quis dizer {spell_checker("mgica")} ?')

Você quis dizer mágica ?


# How do I analyze whether the spell checker is acceptable?

> One way to do this is to run the spell checker in a dictionary where you have the wrong word and the correct one, and check if the spell checker, upon receiving the wrong one, can suggest the correct one.

> Take the total value of this list and check how many the spell checker got right to measure its effectiveness.

## Read the database for test

In [146]:
def generate_test_base():
  list_words_test = []
  target_url = 'https://raw.githubusercontent.com/natelson/python/main/machine_learning/data/words_for_test.txt'
  f = get_text_from_a_url_file(target_url).split('\n')
  for line in f:
    if len(line) > 2:
      word_right, word_wrong = line.split()
      list_words_test.append((word_right, word_wrong))

  return list_words_test

In [147]:
print(generate_test_base())

[('podemos', 'pyodemos'), ('esse', 'esje'), ('já', 'jrá'), ('nosso', 'nossov'), ('são', 'sãêo'), ('dos', 'dosa'), ('muito', 'muifo'), ('imagem', 'iômagem'), ('sua', 'ósua'), ('também', 'tambéùm'), ('ele', 'eme'), ('fazer', 'èazer'), ('temos', 'temfs'), ('essa', 'eàssa'), ('quando', 'quaôdo'), ('vamos', 'vamvos'), ('sobre', 'hsobre'), ('java', 'sjava'), ('das', 'daõs'), ('agora', 'agorah'), ('está', 'eòtá'), ('cada', 'céda'), ('mesmo', 'zmesmo'), ('nos', 'noâ'), ('forma', 'fobma'), ('seja', 'sejéa'), ('então', 'enêão'), ('criar', 'èriar'), ('código', 'cóeigo'), ('caso', 'casío'), ('exemplo', 'áexemplo'), ('tem', 'tĩem'), ('usuário', 'usuárôio'), ('dados', 'dfados'), ('python', 'pgthon'), ('nossa', 'nossah'), ('além', 'alémè'), ('assim', 'asõim'), ('ter', 'teb'), ('até', 'atĩ'), ('bem', 'âem'), ('design', 'desigen'), ('trabalho', 'trabalàho'), ('foi', 'foo'), ('apenas', 'apenaũ'), ('empresa', 'empresà'), ('valor', 'valíor'), ('será', 'serr'), ('entre', 'entke'), ('método', 'méqodo'), ('p

> This first version receives the list of test words, and calls the spell_checker with the wrong word and compares it with what should be the right one and counts the hits and returns the hit percentage.

In [148]:
def spell_checker_correct_percentage(test_base):
  correct_answears = 0
  for word_right, word_wrong in test_base:
    word_test = spell_checker(word_wrong)
    if word_test == word_right:
      correct_answears += 1
  
  percentual = (correct_answears/len(test_base)) * 100

  return round(percentual,2)




In [149]:
test_base = generate_test_base()
vocabulary = list(set(tokens_only_words_unique))

In [150]:
print(f'Accuracity of {spell_checker_correct_percentage(test_base)}%')

Accuracity of 76.34%


> We have a hit rate of 76.34%, but we need to remember that there are two factors that can influence this percentage, the first is our implementation, that is, words that we were unable to identify its variation in our corpus and the second are words that were not taught to our algorithm, so we need to measure the number of unknown words in our database.

In [151]:
def spell_checker_correct_percentage(test_base, vocabulary):
  correct_answears = 0
  unknown_vocabulary = 0
  for word_right, word_wrong in test_base:
    word_test = spell_checker(word_wrong)
    unknown_vocabulary += (word_right not in vocabulary)
    if word_test == word_right:
      correct_answears += 1
    
      
  
  percentual = (correct_answears/len(test_base)) * 100
  unknown_percentual =  (unknown_vocabulary/len(test_base)) * 100

  return round(percentual,2), round(unknown_percentual,2)


In [152]:
percentual, unknown_percentual = spell_checker_correct_percentage(test_base, vocabulary)

In [153]:
print(f'Accuracity of {percentual}% and unknow of {unknown_percentual}%' )

Accuracity of 76.34% and unknow of 6.99%


> Well from our test base, 6.99% of the correct words were not taught to our broker, one way to solve this is to improve our teaching base

## The goal here is not to build the best spell checker, but to explore some NLP functions that would help us build one. Until next time folks.